src/NodeApi/JSReference.cs

```diff
@@ -23,8 +23,9 @@ namespace Microsoft.JavaScript.NodeApi;
 /// </remarks>
 public class JSReference : IDisposable
 {
-    private readonly JSRuntimeContext? _context;
     private readonly napi_ref _handle;
+    private readonly napi_env _env;
```

Comment:
>_env [](http://example.com/codeflow?start=30&length=4)

The new code uses unchecked napi_env. It never verifies that the _env is still valid and not points to a deleted memory. I agree that it could be a bit better than before where we used some random current napi_env, but it is not as safe as always going through the JSRuntimeContext that we had originally. It would be great to restore the use of the JSRuntimeContext. We must not allow JSReference creation without valid JSRuntimeContext.

Reply:
> It never verifies that the _env is still valid

Neither does `JSRuntimeContext`. It checks if `_env` is non-null, but that doesn't guarantee it has not been disposed.
```C#
public static explicit operator napi_env(JSRuntimeContext context) => context?._env ??
    throw new ArgumentNullException(nameof(context));
```

> We must not allow JSReference creation without valid JSRuntimeContext.

`NativeHost` needs to create references but should not establish the module context as that needs to be owned by the `ManagedHost`. That is the purpose of the "no-context" value scope type.

Reply:
Also how is this different from `JSValue` which has a `JSValueScope` field, which again only checks for null when converting to `napi_env`?

Reply:
Then it does not sound right too. `JSValue` used to ensure the safe use of `napi_value` by checking disposed status of the `JSValueScope`. I will check when it is changed. Here we also need a way to ensure that `napi_env` is safe when we use it from the `JSReference`. The initial idea was simple: `napi_env` has instance data. This instance data connected to `JSContext` that was later converted to `JSRuntimeContext`. When `napi_env` is deleted it calls the instance data finalizer that disposes the `JSContext`. The `JSReference` checks the `JSContext` disposed status to make decision if the `napi_env` is still alive. Note that `napi_env` also removes all `napi_ref` when deleted. So, the the wrapped `napi_ref` could also become invalid if we are not aware about `napi_env` deletion.

It is OK to merge this PR "as-is". I will see how to fix it all.

Reply:
> JSValue used to ensure the safe use of napi_value by checking disposed status of the JSValueScope. I will check when it is changed.

It's possible some of my refactoring broke that without realizing it, sorry. Clearly we need tests to validate those kinds of protections. But I think that will be difficult without being able to mock out the JS runtime. So I hope we can come up with a workable approach for that, as discussed yesterday.

Reply:
I opened #121 "Add tests to validate checks for mis-use of native handles"